File: EduPulseBackend/src/main/java/com/example/edupulse_backend/service/NotificationService.java
package com.example.edupulse_backend.service;

import com.example.edupulse_backend.model.Notification;
import com.example.edupulse_backend.payload.response.ResponseDto;

public interface NotificationService {
    
    /**
     * Get all notifications for a user
     */
    ResponseDto getNotifications(String userId);
    
    /**
     * Get only unread notifications for a user
     */
    ResponseDto getUnreadNotifications(String userId);
    
    /**
     * Get count of unread notifications
     */
    ResponseDto getUnreadCount(String userId);
    
    /**
     * Mark a notification as read
     */
    ResponseDto markAsRead(String notificationId);
    
    /**
     * Mark all notifications as read for a user
     */
    ResponseDto markAllAsRead(String userId);
    
    /**
     * Poll for new notifications (for long polling implementation)
     */
    ResponseDto pollNotifications(String userId, Long lastPolledTimestamp);
    
    /**
     * Create and save a new notification
     */
    Notification createNotification(String recipientId, String senderId, String senderName,
                                  Notification.NotificationType type, String content);
    
    /**
     * Create a notification with optional post and comment IDs
     */
    Notification createDetailedNotification(String recipientId, String senderId, String senderName,
                                     String postId, String commentId,
                                     Notification.NotificationType type, String content);
}

File: client/src/services/notificationService.js
import axios from './axiosConfig';

/**
 * Service for interacting with the notifications API
 */
export const notificationService = {
  /**
   * Get all notifications for the current user
   */
  async getNotifications() {
    try {
      const response = await axios.get('/api/v1/notifications');
      return response.data;
    } catch (error) {
      console.error('Error fetching notifications:', error);
      return {
        error: true,
        message: error.response?.data?.message || 'Failed to fetch notifications'
      };
    }
  },

  /**
   * Get only unread notifications for the current user
   */
  async getUnreadNotifications() {
    try {
      const response = await axios.get('/api/v1/notifications/unread');
      return response.data;
    } catch (error) {
      console.error('Error fetching unread notifications:', error);
      return {
        error: true,
        message: error.response?.data?.message || 'Failed to fetch unread notifications'
      };
    }
  },

  /**
   * Get count of unread notifications
   */
  async getUnreadCount() {
    try {
      const response = await axios.get('/api/v1/notifications/count');
      return response.data;
    } catch (error) {
      console.error('Error fetching unread notification count:', error);
      return {
        error: true,
        message: error.response?.data?.message || 'Failed to fetch notification count'
      };
    }
  },

  /**
   * Mark a specific notification as read
   * @param {string} notificationId - ID of the notification to mark as read
   */
  async markAsRead(notificationId) {
    try {
      const response = await axios.put(`/api/v1/notifications/read/${notificationId}`);
      return response.data;
    } catch (error) {
      console.error('Error marking notification as read:', error);
      return {
        error: true,
        message: error.response?.data?.message || 'Failed to mark notification as read'
      };
    }
  },

  /**
   * Mark all notifications as read for the current user
   */
  async markAllAsRead() {
    try {
      const response = await axios.put('/api/v1/notifications/read-all');
      return response.data;
    } catch (error) {
      console.error('Error marking all notifications as read:', error);
      return {
        error: true,
        message: error.response?.data?.message || 'Failed to mark all notifications as read'
      };
    }
  },

  /**
   * Long polling endpoint for real-time notifications
   * @param {number} lastTimestamp - Timestamp of the last check for notifications
   */
  async pollNotifications(lastTimestamp = null) {
    try {
      const url = lastTimestamp 
        ? `/api/v1/notifications/poll?lastTimestamp=${lastTimestamp}` 
        : '/api/v1/notifications/poll';
        
      const response = await axios.get(url);
      return response.data;
    } catch (error) {
      console.error('Error polling for notifications:', error);
      return {
        error: true,
        message: error.response?.data?.message || 'Failed to poll for notifications'
      };
    }
  }
};